### PR TITLE
Batch updates second try

### PR DESCRIPTION
**Pull request by Szymon20000**
### The Problem this pull-request solves
One of the main problems with react-native-onyx is that it renders every component in a separate micro-task. Because right now we don't use concurrent rendering every setState not within the same event listener will cause a separate render. Every react render has certain overhead so if it's possible to render 2 state changes together we will only pay for the overhead once. Additionally we introduce additional overhead in render side-effects. For instance in SidebarLinks component we compute which report should be visible. That method is heavy and we run it almost on every render. If we batch 2 setStates together that would save us one such computation. This pull-requests batches all setStates caused by onyx changes that occurs within the same microtask cycle. That saves a ton of time. 

#### Examples of the problem
1) Imagine there are 10 components that are listening to a current report changes. Currently all those components will be notified separately 
2) Imagine we have one component that listens to 4 different onyx props that are changed one by one (SidebarLinks is such component) right now we will render is as many times as we have changes. With this pr we will merge it only once per micotask cycle what will avoid a lot of recomputation

> Note that is change won't be needed anymore once we use concurrent rendering. 

### Details
We use  `unstable_batchedUpdates` method from react (That is currently used in even listeners so it's pretty safe). 
This method triggers single react render for all the setState changes passed to it. 

#### How it works 
Screenshot 2023-08-28 at 14 46 16
For every notifySubscribers or collection subscribers we don't queue the update on micro task queue instead we add that change to pending changes list that is flushed at the end of the current micro task cycle. We do it by setting macro task which is `setTimeout(() => {}, 0)`.
### Related Issues

GH_LINK

### CODE REVIEW

**Comment by Li357** [comment]
Is this ready for review?

File: lib/Onyx.js [hunk 1]
```diff
@@ -801,22 +803,58 @@ function disconnect(connectionID, keyToRemoveFromEvictionBlocklist) {
...
+    scheduledPromise = new Promise((resolve) => {
+        setTimeout(() => {
+            scheduledPromise = null;
+            const updatesCopy = pendingUpdates;
+            pendingUpdates = [];
+            const updatesCollectionCopy = pendingCollectionUpdates;
+            pendingCollectionUpdates = [];
+            unstable_batchedUpdates(() => {
+                for (let i = 0; i < updatesCopy.length; ++i) {
+                    keyChanged(updatesCopy[i][0], updatesCopy[i][1], updatesCopy[i][2]);
```
Comment (Julesssss) [inline lib/Onyx.js:832]: I couldn't find anything that confirms that the `pendingUpdate|updatesCopy` is has a 3rd child. Presumably, this is guaranteed as that's how our Onyx updates are written?

**Review by Julesssss — APPROVED** [review]
While this looks good, it would be great to test the component renders to prove the updates are now batched.

File: jestSetup.js [hunk 2]
```diff
@@ -2,7 +2,10 @@ jest.mock('./lib/storage');
 jest.mock('./lib/storage/NativeStorage', () => require('./lib/storage/__mocks__'));
 jest.mock('./lib/storage/WebStorage', () => require('./lib/storage/__mocks__'));
 jest.mock('./lib/storage/providers/IDBKeyVal', () => require('./lib/storage/__mocks__'));
+
 jest.mock('react-native-device-info', () => ({getFreeDiskStorage: () => {}}));
 jest.mock('react-native-quick-sqlite', () => ({
     open: () => ({execute: () => {}}),
 }));
+
+jest.useRealTimers();
```
Comment (tgolen) [inline jestSetup.js:11]: Why is this added here? Do you mind adding a code comment to explain?

File: lib/Onyx.js [hunk 3]
```diff
@@ -801,22 +803,58 @@ function disconnect(connectionID, keyToRemoveFromEvictionBlocklist) {
     delete callbackToStateMapping[connectionID];
 }
 
+// [key, value, canUpdateSubscriber]
+let pendingUpdates = [];
+let scheduledPromise;
+let pendingCollectionUpdates = [];
```
Comment (tgolen) [inline lib/Onyx.js:809]: Could you also please add a doc comment for this (like you did for `pendingUpdates`)?

File: lib/Onyx.js [hunk 4]
```diff
@@ -801,22 +803,58 @@ function disconnect(connectionID, keyToRemoveFromEvictionBlocklist) {
...
+ * cause react to schedule the updates at once instead of after each other. This is mainly a performance optimization.
+ * @returns {Promise}
+ */
+function scheduleOnyxFlushIfNeeded() {
+    if (scheduledPromise) {
+        return scheduledPromise;
+    }
+
+    scheduledPromise = new Promise((resolve) => {
+        setTimeout(() => {
```
Comment (tgolen) [inline lib/Onyx.js:824]: Please add a code comment explaining why it's necessary to have a `setTimout(0)` here.

File: lib/Onyx.js [hunk 5]
```diff
@@ -20,6 +21,7 @@ const METHOD = {
 
 // Key/value store of Onyx key and arrays of values to merge
 const mergeQueue = {};
+const mergeQueuePromise = {};
```
Comment (tgolen) [inline lib/Onyx.js:24]: I followed most of the changes in this PR except for where it came to `mergeQueuePromise`. Could you add some code comments to give a general idea around what it's for and how it's used?

**Review by tgolen — CHANGES_REQUESTED** [review]
Is there any way to write some unit tests to verify that the updates were batched as expected? Something like... have a mocked function that is triggered on render, and updating multiple properties that the component is listening to triggers the render callback the right number of times.

This is an example test of what I'm thinking of: https://github.com/Expensify/react-native-onyx/blob/ad5533e199b348713f7ff9dd4b75894af7a1a597/tests/unit/withOnyxTest.js#L59-L77

File: lib/Onyx.js [hunk 6]
```diff
@@ -966,15 +1006,16 @@ function set(key, value) {
     const hasChanged = cache.hasValueChanged(key, valueWithNullRemoved);
 
     // This approach prioritizes fast UI changes without waiting for data to be stored in device storage.
-    broadcastUpdate(key, valueWithNullRemoved, hasChanged, 'set');
+    const updatePromise = broadcastUpdate(key, valueWithNullRemoved, hasChanged, 'set');
 
     // If the value has not changed, calling Storage.setItem() would be redundant and a waste of performance, so return early instead.
     if (!hasChanged) {
-        return Promise.resolve();
+        return updatePromise;
```
Comment (marcaaron) [inline lib/Onyx.js:1013]: NAB, I am a _little_ skeptical of this change mainly because it is now harder to reason about when this promise will resolve. 

Before we can easily see it will either resolve right away or after the value is saved to storage. Now it's less clear... it's still both of those, but maybe also however long it takes the queue to flush? Ultimately, I don't know if this concern is valid or if it makes any difference to the existing code we have.

File: lib/Onyx.js [hunk 7]
```diff
@@ -1091,20 +1133,22 @@ function merge(key, changes) {
...
-                    return Promise.resolve();
+                    return updatePromise;
                 }
 
-                return Storage.mergeItem(key, batchedChanges, modifiedData);
+                return Storage.mergeItem(key, batchedChanges, modifiedData)
+                    .then(() => updatePromise);
             } catch (error) {
                 Logger.logAlert(`An error occurred while applying merge for key: ${key}, Error: ${error}`);
+                return Promise.resolve();
```
Comment (marcaaron) [inline lib/Onyx.js:1147]: well, this change actually will have some interesting consequences (I _think_ probably good ones) when a storage error happens.

e.g. if something breaks like not able to save the merge to storage because you are out of space then we will resolve the promise (which would allow anything depending on it to complete to continue as well - which sounds good).

But, if I am not mistaken - I think in this case... we can and should still wait for the `updatePromise` to resolve? 🤔 

Unless that throws an error too. And in that case we should return the resolved promise 😄

**Review by marcaaron — CHANGES_REQUESTED** [review]

**Comment by marcaaron** [comment]
>One of the main problems with react-native-onyx is that it renders every component in a seperate promise

I am a little unsure about what “renders every component in a separate promise” means exactly. Could you expand on that a bit?

>right now we don't use concurrent rendering every setState not within the same event listener will cause a separate render

Is this referring to what happens in the `notifySubscribersOnNextTick()` method?

https://github.com/Expensify/react-native-onyx/blob/7f23ec8a279c93616b9e2d43ccb1574ccd6be9c9/lib/Onyx.js#L805-L820

One thought I had while reading the comment is if we still really _need_ that promise. It seems unnecessary and I wonder if it really is dangerous to remove - maybe it just needs to be a “breaking change” and we fix whatever might depend on this logic in App.

Though, thinking more on that - we will always have the `get()` in `Onyx.merge()` that would create a promise. So, this observation doesn't affect the solution you are proposing much.

File: jestSetup.js (hunk 2 above)
Comment (Szymon20000) [reply jestSetup.js:11]: Hanno added that because otherwise setTimeout is not executed unless we call jest.advanceTime(). So to avoid calling that everywhere we used real times.

File: lib/Onyx.js (hunk 4 above)
Comment (Szymon20000) [reply lib/Onyx.js:829]: I added a comment explaining when the batch will be flushed. Explaining how timers and batching works in react native is not possible in a single comment so I simplified it a bit.

File: lib/Onyx.js (hunk 5 above)
Comment (Szymon20000) [reply lib/Onyx.js:24]: This change is mostly for adjusting tests from what I remember. 
Looks like we use it in tests for making sure all the pending merges has been applied.

File: lib/Onyx.js [hunk 8]
```diff
@@ -801,22 +803,58 @@ function disconnect(connectionID, keyToRemoveFromEvictionBlocklist) {
...
+let pendingUpdates = [];
+let scheduledPromise;
+let pendingCollectionUpdates = [];
+
+/**
+ * We are batching together onyx updates. This helps with use cases where we schedule onyx updates after each other.
+ * This happens for example in the Onyx.update function, where we process API responses that might contain a lot of
+ * update operations. Instead of calling the subscribers for each update operation, we batch them together which will
+ * cause react to schedule the updates at once instead of after each other. This is mainly a performance optimization.
+ * @returns {Promise}
```
[the remark this answers is not in the document]
Comment (Szymon20000) [reply lib/Onyx.js:817]: Hanno already renamed all variables and functions so they use "subscriberUpdates".

File: lib/Onyx.js (hunk 1 above)
Comment (Szymon20000) [reply lib/Onyx.js:837]: Screenshot 2023-09-05 at 09 39 03
The keyChanged method has 3 arguments so we only delay the execution.

File: lib/Onyx.js [hunk 9]
```diff
@@ -801,22 +803,58 @@ function disconnect(connectionID, keyToRemoveFromEvictionBlocklist) {
...
+        }, 0);
+    });
+    return scheduledPromise;
+}
+
 /**
- * This method mostly exists for historical reasons as this library was initially designed without a memory cache and one was added later.
- * For this reason, Onyx works more similar to what you might expect from a native AsyncStorage with reads, writes, etc all becoming
- * available async. Since we have code in our main applications that might expect things to work this way it's not safe to change this
- * behavior just yet.
```
[the remark this answers is not in the document]
Comment (Szymon20000) [reply lib/Onyx.js:808]: Changed!

File: lib/Onyx.js (hunk 7 above)
Comment (Szymon20000) [reply lib/Onyx.js:1152]: I don't see the difference to be honest :) Looks like Promise.resolve was also returned in this case just right below catch block.

**Comment by Szymon20000** [comment]
@marcaaron As long as we don't use fabric we are not able to use concurrent rendering. As a result whenever you call setState on a component it will right away cause a synchronous render (apart from few edge cases like events or active render). Because onyx notifies withOnyx components in separate micro-tasks it causes all of them to start synchronous render.  By “renders every component in a separate promise” I meant separate micro tasks :)

File: lib/Onyx.js [hunk 10]
```diff
@@ -801,22 +803,58 @@ function disconnect(connectionID, keyToRemoveFromEvictionBlocklist) {
     delete callbackToStateMapping[connectionID];
 }
 
+// [key, value, canUpdateSubscriber]
```
[the remark this answers is not in the document]
Comment (Szymon20000) [reply lib/Onyx.js]: It will have the same types as keyChange and keysChanged method and will be typed once we rewrite the code to typescript. Can we leave it for now?

**Comment by Szymon20000** [comment]
Found how it can be done

File: tests/unit/withOnyxTest.js [hunk 11]
```diff
@@ -76,6 +77,34 @@ describe('withOnyx', () => {
...
+        render(<TestComponentWithOnyx onRender={onRender} />);
+
+        return waitForPromisesToResolve()
+            .then(() => {
+                Onyx.mergeCollection(ONYX_KEYS.COLLECTION.TEST_KEY, {test_1: {ID: 123}});
+                Onyx.merge(ONYX_KEYS.SIMPLE_KEY, 'string');
+                return Onyx.merge(ONYX_KEYS.SIMPLE_KEY_2, 'string2');
+            })
+            .then(() => {
+                expect(onRender).toHaveBeenCalledTimes(2);
```
Comment (tgolen) [inline tests/unit/withOnyxTest.js:104]: Could you please add a comment to explain why it's expected to be called exactly two times?

Comment (Szymon20000) [reply tests/unit/withOnyxTest.js:104]: Sure, just a sec

Comment (Szymon20000) [reply tests/unit/withOnyxTest.js:106]: done

File: lib/Onyx.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (marcaaron) [reply lib/Onyx.js:819]: Cool thanks!

File: lib/Onyx.js [hunk 12]
```diff
@@ -801,22 +803,65 @@ function disconnect(connectionID, keyToRemoveFromEvictionBlocklist) {
...
+ * cause react to schedule the updates at once instead of after each other. This is mainly a performance optimization.
+ * @returns {Promise}
+ */
+function scheduleSubscriberUpdatesFlushIfNeeded() {
+    if (scheduledPromise) {
+        return scheduledPromise;
+    }
+
+    scheduledPromise = new Promise((resolve) => {
+        /* We use (setTimout, 0) here which should be called once native module calles are flushed (usually at the end of the frame)
```
Comment (marcaaron) [inline lib/Onyx.js:827]:
```suggestion
        /* We use (setTimeout, 0) here which should be called once native module calls are flushed (usually at the end of the frame)
```

File: lib/Onyx.js [hunk 13]
```diff
@@ -801,22 +803,65 @@ function disconnect(connectionID, keyToRemoveFromEvictionBlocklist) {
...
+         * We may investigate if (setTimout, 1) (which in React Native is equal to requestAnimationFrame) works even better
```
Comment (marcaaron) [inline lib/Onyx.js:828]:
```suggestion
         * We may investigate if (setTimeout, 1) (which in React Native is equal to requestAnimationFrame) works even better
```

File: lib/Onyx.js (hunk 7 above)
Comment (marcaaron) [reply lib/Onyx.js:1154]: Oh duh, ok I missed that line I guess 😄

**Review by marcaaron — CHANGES_REQUESTED** [review]
Just small typos to fix now. I will resolve my NABs.

File: lib/Onyx.js (hunk 6 above)
Comment (marcaaron) [reply lib/Onyx.js:1020]: we will find out I guess 😄

**Comment by Szymon20000** [comment]
The App tests are failing with this pr. I will let you know when they are fixed. 
I just realised that we only want to batch withOnyx subscribers as regular ones should be notified right away.

**Comment by Szymon20000** [comment]
@tgolen @marcaaron Are there any places in the app when we trigger Onyx update in Onyx.connect subscriber? 
If that's the case then I think we shouldn't batch them. 
I can image that we have places like that when once onyx.update triggers another one. 
So for examples. update A -> update B -> update C would require 3 batches now. 
If you think we may have such case in the app I create a pr that fixes that problem. https://github.com/margelo/react-native-onyx/pull/1

**Comment by tgolen** [comment]
I just looked at all the usages of `Onyx.update` and there are none that happen inside a `connect()` subscriber.

**Comment by marcaaron** [comment]
>I just realised that we only want to batch withOnyx subscribers as regular ones should be notified right away

Oh, right, because they do not call `setState()`. Makes sense. 

>Are there any places in the app when we trigger Onyx update in Onyx.connect subscriber?
>If that's the case then I think we shouldn't batch them.

Sorry, I did not follow the reasoning. Is there some kind of critical problem that we will see if we batch those updates?

**Comment by Szymon20000** [comment]
Ah if we don't change state in onyx connect then probably it doesn't make any difference. I did it because I thought you may have some reactive code that is triggered once we merge something.

**Comment by Szymon20000** [comment]
Screenshot 2023-09-07 at 16 03 19
@tgolen is may be the case with network calls

**Comment by tgolen** [comment]
Ah, you're right. I'm on a branch where we've refactored that specific code to be a little different and it's not clear anymore that the `Onyx.update()` is inside the `connect()`.

https://github.com/Expensify/App/blob/5373041c5776fd8a51ea3646710ed9c7ccc50658/src/libs/Network/SequentialQueue.js#L126

**Comment by Szymon20000** [comment]
I was able to fix most of failing tests in the app repo. There are 6 that left. 
I think we need the pr for only batching withOnyx subscribers as networking flushing would be delayed otherwise.

**Review by tgolen — APPROVED** [review]

**Review by marcaaron — APPROVED** [review]

**Review by tgolen — APPROVED** [review]

**Comment by mountiny** [comment]
To get this going ahead I will merge this now given there are two approvals here

**Comment by marcaaron** [comment]
Should we make an announcement not to update the Onyx version until that testing is done? These feel like pretty significant changes.
